File: fastify-ts-bun-backend/tests/error-handling.test.ts
import Fastify, { type FastifyInstance } from 'fastify';
import { afterEach, describe, expect, it } from 'vitest';
import { setFastifyErrorHandler } from '../src/middleware/error-handler';
import { setFastifyNotFoundHandler } from '../src/middleware/not-found';

describe('error handling', () => {
  let app: FastifyInstance | undefined;

  afterEach(async () => {
    if (app) {
      await app.close();
      app = undefined;
    }
  });

  it('returns 404 for unknown routes', async () => {
    app = Fastify({ logger: false });
    setFastifyNotFoundHandler(app);
    setFastifyErrorHandler(app);

    const response = await app.inject({
      method: 'GET',
      url: '/missing',
    });

    expect(response.statusCode).toBe(404);
    const body = response.json();
    expect(body.status).toBe(404);
    expect(body.code).toBe('NOT_FOUND');
    expect(body.title).toBe('Not Found');
    expect(body.type).toContain('/not-found');
    expect(body.requestId).toBeDefined();
  });

  it('returns normalized 500 payload', async () => {
    app = Fastify({ logger: false });

    app.get('/boom', async () => {
      throw new Error('boom');
    });

    setFastifyErrorHandler(app);

    const response = await app.inject({
      method: 'GET',
      url: '/boom',
    });

    expect(response.statusCode).toBe(500);
    const body = response.json();
    expect(body.status).toBe(500);
    expect(body.code).toBe('INTERNAL_SERVER_ERROR');
    expect(body.detail).toBe('Unexpected server error.');
    expect(body.title).toBe('Internal Server Error');
    expect(body.requestId).toBeDefined();
  });
});

File: flask-python-backend/tests/error_handling_test.py
from src.app import create_app


def test_returns_404_for_unknown_routes(client) -> None:
    response = client.get('/missing')
    payload = response.get_json()

    assert response.status_code == 404
    assert payload['status'] == 404
    assert payload['code'] == 'NOT_FOUND'
    assert payload['title'] == 'Not Found'
    assert '/not-found' in payload['type']


def test_returns_normalized_500_payload() -> None:
    app = create_app()

    @app.get('/boom')
    def boom() -> None:
        raise RuntimeError('boom')

    with app.test_client() as client:
        response = client.get('/boom')

    payload = response.get_json()
    assert response.status_code == 500
    assert payload['status'] == 500
    assert payload['code'] == 'INTERNAL_SERVER_ERROR'
    assert payload['detail'] == 'Unexpected server error.'


def test_rate_limit_eventually_returns_429(client) -> None:
    status_codes = [client.get('/api/v1/ping').status_code for _ in range(105)]

    assert 429 in status_codes
